### Customer churn EDA

Import pandas and numpy

In [1]:
import pandas as pd
import numpy as np

Read the data and take a look at the first 5 lines of the dataframe

In [4]:
df = pd.read_csv('data/telecom_churn.csv')
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

What is dimentionality of our data?

In [5]:
df.shape

(3333, 20)

What are the feature names?

In [6]:
list(df.columns)

['State',
 'Account length',
 'Area code',
 'International plan',
 'Voice mail plan',
 'Number vmail messages',
 'Total day minutes',
 'Total day calls',
 'Total day charge',
 'Total eve minutes',
 'Total eve calls',
 'Total eve charge',
 'Total night minutes',
 'Total night calls',
 'Total night charge',
 'Total intl minutes',
 'Total intl calls',
 'Total intl charge',
 'Customer service calls',
 'Churn']

Get general information of dataframe. 

In [153]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

Change dtype of feature "Churn" from boolean to integer.

In [7]:
df['Churn'] = df['Churn'].astype(int)

Show basic statistical characteristics of numeric features. (Number of non missing values, mean, standard deviation, median , etc)

In [8]:
df.describe()

Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       Total eve charge  Total night minutes  Total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       Total intl charge  Customer service calls        Churn  
count        3333.000000             3333.000000  3333.000000  
mean            2.764581                1.562856     0.144914  
std             0.753773                1.315491     0.352067  
min             0.000000                0.000000     0.000000  
25%             2.300000                1.000000     0.000000  
50%             2.780000                1.000000     0.000000  
75%             3.270000                2.000000     0.000000  
max             5.400000                9.000000     1.000000

In [9]:
df.isna().sum()

State                     0
Account length            0
Area code                 0
International plan        0
Voice mail plan           0
Number vmail messages     0
Total day minutes         0
Total day calls           0
Total day charge          0
Total eve minutes         0
Total eve calls           0
Total eve charge          0
Total night minutes       0
Total night calls         0
Total night charge        0
Total intl minutes        0
Total intl calls          0
Total intl charge         0
Customer service calls    0
Churn                     0
dtype: int64

Show statistics on non-numerical features, using the same method as above. 

In [10]:
df.describe(include='O')

State International plan Voice mail plan
count   3333               3333            3333
unique    51                  2               2
top       WV                 No              No
freq     106               3010            2411

Show count of values of "Churn" feature

In [13]:
df.Churn.value_counts()

0    2850
1     483
Name: Churn, dtype: int64

Show normalized counts of values for "Churn" 

In [14]:
df.Churn.value_counts(normalize=True)

0    0.855086
1    0.144914
Name: Churn, dtype: float64


### Sorting

Sort by *Total day charge* (sort in descending order):

In [16]:
df.sort_values(by="Total day charge", ascending=False).head()

State  Account length  Area code International plan Voice mail plan  \
365     CO             154        415                 No              No   
985     NY              64        415                Yes              No   
2594    OH             115        510                Yes              No   
156     OH              83        415                 No              No   
605     MO             112        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
365                       0              350.8               75   
985                       0              346.8               55   
2594                      0              345.3               81   
156                       0              337.4              120   
605                       0              335.5               77   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
365              59.64              216.5               94             18.40   
985              58.96              249.5               79             21.21   
2594             58.70              203.4              106             17.29   
156              57.36              227.4              116             19.33   
605              57.04              212.5              109             18.06   

      Total night minutes  Total night calls  Total night charge  \
365                 253.9                100               11.43   
985                 275.4                102               12.39   
2594                217.5                107                9.79   
156                 153.9                114                6.93   
605                 265.0                132               11.93   

      Total intl minutes  Total intl calls  Total intl charge  \
365                 10.1                 9               2.73   
985                 13.3                 9               3.59   
2594                11.8                 8               3.19   
156                 15.8                 7               4.27   
605                 12.7                 8               3.43   

      Customer service calls  Churn  
365                        1      1  
985                        1      1  
2594                       1      1  
156                        0      1  
605                        2      1

What is the proportion of churned users in our dataframe?

In [17]:
df.Churn.value_counts(1)[1] 

0.14491449144914492

14.5% is actually quite bad for a company; such a churn rate can make the company go bankrupt.

**What are average values of numerical features for churned users?**

In [18]:
df[df["Churn"] == 1].mean()

/var/folders/c2/lxpvl7kx34qg2cwgv5hmsv7c0000gn/T/ipykernel_70411/507517496.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df[df["Churn"] == 1].mean()


Account length            102.664596
Area code                 437.817805
Number vmail messages       5.115942
Total day minutes         206.914079
Total day calls           101.335404
Total day charge           35.175921
Total eve minutes         212.410145
Total eve calls           100.561077
Total eve charge           18.054969
Total night minutes       205.231677
Total night calls         100.399586
Total night charge          9.235528
Total intl minutes         10.700000
Total intl calls            4.163561
Total intl charge           2.889545
Customer service calls      2.229814
Churn                       1.000000
dtype: float64

**How much time (on average) do churned users spend on the phone during daytime (it's "Total day minutes" column)?**

In [19]:
df[df['Churn'] == 1].describe().loc[['mean'], ["Total day minutes"]]

Total day minutes
mean         206.914079


**What is the maximum length of international calls(Total intl minutes) among loyal users (not churned) who do not have an international plan?**

In [20]:
df[(df['Churn'] == 0) & (df['International plan'] == 'No')]['Total intl minutes'].max()

18.9

**DataFrames can be indexed by column name (label) or row name (index) or by the serial number of a row. The `loc` method is used for indexing by name, while `iloc()` is used for indexing by number.**

Show values of the rows with index from 0 to 5 (inclusive) and columns labeled from State to Area code (inclusive).

In [24]:
df.loc[(range(6)),'State':'Area code']

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415
5    AL             118        510

In [25]:
#df.loc[0:5, "State":"Area code"]

Show the values of the first five rows in the first three columns" (as in a typical Python slice).

In [26]:
df.iloc[:5, :3]

State  Account length  Area code
365     CO             154        415
985     NY              64        415
2594    OH             115        510
156     OH              83        415
605     MO             112        415

Show the last line of the dataframe with slicing

In [30]:
df[-1:]

State  Account length  Area code International plan Voice mail plan  \
1345    SD              98        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
1345                      0                0.0                0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
1345               0.0              159.6              130             13.57   

      Total night minutes  Total night calls  Total night charge  \
1345                167.1                 88                7.52   

      Total intl minutes  Total intl calls  Total intl charge  \
1345                 6.8                 1               1.84   

      Customer service calls  Churn  
1345                       4      1


### Applying Functions to Cells, Columns and Rows



### If you want to get the maximum value for each column in a dataframe you can do this

In [31]:
df.apply(np.max)

State                        WY
Account length              243
Area code                   510
International plan          Yes
Voice mail plan             Yes
Number vmail messages        51
Total day minutes         350.8
Total day calls             165
Total day charge          59.64
Total eve minutes         363.7
Total eve calls             170
Total eve charge          30.91
Total night minutes       395.0
Total night calls           175
Total night charge        17.77
Total intl minutes         20.0
Total intl calls             20
Total intl charge           5.4
Customer service calls        9
Churn                         1
dtype: object

Show the part of dataframe where "State" starts with "W". Use apply function with lambda inside it.

In [32]:
df[df['State'].apply(lambda x: x[0] == 'W')]

State  Account length  Area code International plan Voice mail plan  \
2412    WI              89        415                 No              No   
2573    WV             124        415                 No              No   
2874    WY             134        510                 No              No   
3268    WA              84        415                 No              No   
1350    WA              55        408                 No              No   
...    ...             ...        ...                ...             ...   
1538    WV             113        415                 No              No   
2859    WV              87        415                 No              No   
2241    WA              56        415                 No             Yes   
1076    WY              53        415                 No             Yes   
1986    WI              70        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
2412                      0              326.3              112   
2573                      0              312.0              112   
2874                      0              296.0               93   
3268                      0              289.1              100   
1350                      0              285.7              124   
...                     ...                ...              ...   
1538                      0               61.2              111   
2859                      0               58.0              125   
2241                     29               37.7              115   
1076                     27               25.9              119   
1986                      0                7.9              100   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
2412             55.47              165.1              110             14.03   
2573             53.04              180.0              109             15.30   
2874             50.32              226.4              117             19.24   
3268             49.15              233.8               97             19.87   
1350             48.57              230.9              106             19.63   
...                ...                ...              ...               ...   
1538             10.40               92.3               88              7.85   
2859              9.86               67.5              116              5.74   
2241              6.41              144.1              111             12.25   
1076              4.40              206.5               96             17.55   
1986              1.34              136.4               83             11.59   

      Total night minutes  Total night calls  Total night charge  \
2412                162.9                 97                7.33   
2573                168.6                 94                7.59   
2874                246.8                 98               11.11   
3268                223.5                148               10.06   
1350                230.7                140               10.38   
...                   ...                ...                 ...   
1538                197.4                114                8.88   
2859                185.9                136                8.37   
2241                226.6                101               10.20   
1076                228.1                 64               10.26   
1986                156.6                 89                7.05   

      Total intl minutes  Total intl calls  Total intl charge  \
2412                 7.5                 1               2.03   
2573                12.8                 4               3.46   
2874                12.3                10               3.32   
3268                12.7                 2               3.43   
1350                14.8                 7               4.00   
...                  ...               ...                ...   
1538                13.7                 3            


### Grouping

As you know, grouping data in Pandas works as follows:


In [33]:
columns_to_show = ["Total day minutes", "Total eve minutes", "Total night minutes"]

df.groupby(["Churn"])[columns_to_show].describe() 

Total day minutes                                                      \
                  count        mean        std  min      25%    50%     75%   
Churn                                                                         
0                2850.0  175.175754  50.181655  0.0  142.825  177.2  210.30   
1                 483.0  206.914079  68.997792  0.0  153.250  217.6  265.95   

             Total eve minutes              ...                 \
         max             count        mean  ...     75%    max   
Churn                                       ...                  
0      315.6            2850.0  199.043298  ...  233.20  361.8   
1      350.8             483.0  212.410145  ...  249.45  363.7   

      Total night minutes                                               \
                    count        mean        std   min     25%     50%   
Churn                                                                    
0                  2850.0  200.133193  51.105032  23.2  165.90  200.25   
1                   483.0  205.231677  47.132825  47.4  171.25  204.80   

                      
          75%    max  
Churn                 
0      234.90  395.0  
1      239.85  354.9  

[2 rows x 24 columns]

Let’s do the same thing, but slightly differently by passing a list of functions to `agg()`:

In [36]:
columns_to_show = ["Total day minutes", "Total eve minutes", "Total night minutes"]

df.groupby(["Churn"])[columns_to_show].agg([np.mean, np.std, np.min, np.max])

Total day minutes                        Total eve minutes             \
                   mean        std amin   amax              mean        std   
Churn                                                                         
0            175.175754  50.181655  0.0  315.6        199.043298  50.292175   
1            206.914079  68.997792  0.0  350.8        212.410145  51.728910   

                   Total night minutes                          
       amin   amax                mean        std  amin   amax  
Churn                                                           
0       0.0  361.8          200.133193  51.105032  23.2  395.0  
1      70.9  363.7          205.231677  47.132825  47.4  354.9

In [37]:
df.groupby(["Churn"])[columns_to_show].agg(["mean", "std","min", "max"])

Total day minutes                        Total eve minutes             \
                   mean        std  min    max              mean        std   
Churn                                                                         
0            175.175754  50.181655  0.0  315.6        199.043298  50.292175   
1            206.914079  68.997792  0.0  350.8        212.410145  51.728910   

                   Total night minutes                          
        min    max                mean        std   min    max  
Churn                                                           
0       0.0  361.8          200.133193  51.105032  23.2  395.0  
1      70.9  363.7          205.231677  47.132825  47.4  354.9


### DataFrame transformations

Calculate the total number of calls for all users. Create the `total_calls` Series. (total calls is total day/eve/night/intl calls combined)

In [38]:
cols = ['Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge']

In [39]:
df['total_calls'] = df[cols].apply(np.sum, axis=1)
df.head()

State  Account length  Area code International plan Voice mail plan  \
365     CO             154        415                 No              No   
985     NY              64        415                Yes              No   
2594    OH             115        510                Yes              No   
156     OH              83        415                 No              No   
605     MO             112        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
365                       0              350.8               75   
985                       0              346.8               55   
2594                      0              345.3               81   
156                       0              337.4              120   
605                       0              335.5               77   

      Total day charge  Total eve minutes  ...  Total eve charge  \
365              59.64              216.5  ...             18.40   
985              58.96              249.5  ...             21.21   
2594             58.70              203.4  ...             17.29   
156              57.36              227.4  ...             19.33   
605              57.04              212.5  ...             18.06   

      Total night minutes  Total night calls  Total night charge  \
365                 253.9                100               11.43   
985                 275.4                102               12.39   
2594                217.5                107                9.79   
156                 153.9                114                6.93   
605                 265.0                132               11.93   

      Total intl minutes  Total intl calls  Total intl charge  \
365                 10.1                 9               2.73   
985                 13.3                 9               3.59   
2594                11.8                 8               3.19   
156                 15.8                 7               4.27   
605                 12.7                 8               3.43   

      Customer service calls  Churn  total_calls  
365                        1      1      1201.50  
985                        1      1      1226.15  
2594                       1      1      1168.97  
156                        0      1      1179.39  
605                        2      1      1242.16  

[5 rows x 21 columns]

Delete the column that you just created.  (sepakan aracy jurn enq qcum ....)

In [40]:
df.drop('total_calls', axis=1, inplace=True)

Delete the rows at indices 1 and 2.

In [41]:
df.drop([1, 2], inplace=True)

In [42]:
df.sort_index()

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
5       AL             118        510                Yes              No   
6       MA             121        510                 No             Yes   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1              110   
3                         0              299.4               71   
4                         0              166.7              113   
5                         0              223.4               98   
6                        24              218.2               88   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4               99             16.78   
3                50.90               61.9               88              5.26   
4                28.34              148.3              122             12.61   
5                37.98              220.6              101             18.75   
6                37.09              348.5              108             29.62   
...                ...                ...              ...               ...   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7                 91               11.01   
3                   196.9                 89                8.86   
4                   186.9                121                8.41   
5                   203.9                118                9.18   
6                   212.6                118                9.57   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
5                    6.3                 6               1.70   
6                    7.5                 7               2.03   
...                  ...               ...                ...   
3328                 9.9                 6            

### The End :) 